rclpy/rclpy/node.py

```diff
@@ -166,8 +170,23 @@ def set_parameters_atomically(self, parameter_list):
             result = SetParametersResult(successful=True)
 
         if result.successful:
+            parameter_event = ParameterEvent()
             for param in parameter_list:
-                self._parameters[param.name] = param
+                if Parameter.Type.NOT_SET == param.type_:
+                    if Parameter.Type.NOT_SET != self.get_parameter(param.name).type_:
+                        # Parameter deleted. (Parameter had value and new value is not set)
+                        parameter_event.deleted_parameters.append(param.get_rcl_parameter())
+                        del self._parameters[param.name]
```

Comment:
What happens in the else case?

(I didnt try it but by reading the code I wonder)

As a user does that mean that if I:
- run list parameters: see parameter `foo`
- send a request to delete `foo` and receive success
- I run list_parameters again:
  - if the parameter was of any type but "NOT_SET": it has been deleted and is not part of the list
  - if the parameter was of type "NOT_SET": it will still be in the list
    - to actually delete it I should:
      - set a value
      - request a deletion

Is that correct?

Reply:
That's correct but the final subcase "if the parameter was of type "NOT_SET": it will still be in the list" is somewhat vacuous as in order for a NOT_SET parameter to get in the list it would need to be added directly to the parameters map circumventing our API.

Reply:
What does vacuous mean?
Do you mean that scenario should not happen and should be handled as such (warning, error, something else than just ignoring it?) ?

Reply:
> What does vacuous mean?

If they use the node parameters API, there won't be a Parameter of type NOT_SET stored and thus it wouldn't be enumerated in a list_parameters service call.

> Do you mean that scenario should not happen and should be handled as such (warning, error, something else than just ignoring it?) ?

The scenario shouldn't happen but I also don't know that it needs to be explicitly warned or errored any more than any other behavior that would arise from manual manipulation of Node class internals. If you're concerned about seeing NOT_SET parameters in the list_parameters results I think an explicit check in that service handler would be the way to achieve it.

All that said, if there's an IllegalStateException or similar in python that we want to raise here I'm not opposed.